File: src/dataset-normalization.ts
import * as fs from 'fs';
import * as path from 'path';

export interface PreparedDataset {
  dataset: string;
  extensions?: string[];
  diagnostics: string[];
}

interface DatasetFile {
  absolutePath: string;
  relativePath: string;
  extension: string;
}

const SUPPORTED_DISCOVERY_EXTENSIONS = new Set([
  'csv',
  'tsv',
  'json',
  'jsonl',
  'xlsx',
  'xls',
  'docx',
  'pdf',
  'pptx',
  'txt',
  'md',
]);

/**
 * Some source exports use newline-delimited JSON records but keep a .json
 * extension. EvalGen treats .json as a single JSON document, and the bundled
 * enhancer treats .json/.jsonl as document-like content. Convert only this
 * disguised JSONL shape to CSV so tabular business records stay tabular.
 */
export function prepareDatasetForWorkflow(
  datasetPath: string,
  workspace: string,
  extensions?: string[],
): PreparedDataset {
  const extensionFilter = extensions && extensions.length > 0
    ? new Set(extensions.map(normalizeExtension).filter(Boolean))
    : undefined;
  const files = collectDatasetFiles(datasetPath, extensionFilter);
  const jsonlJsonFiles = files.filter((file) => file.extension === 'json' && isJsonLinesFile(file.absolutePath));

  if (jsonlJsonFiles.length === 0) {
    return { dataset: datasetPath, extensions, diagnostics: [] };
  }

  const stageDir = path.join(workspace, '00-normalized-dataset');
  fs.rmSync(stageDir, { recursive: true, force: true });
  fs.mkdirSync(stageDir, { recursive: true });

  const effectiveExtensions = new Set<string>();
  let convertedRows = 0;
  const convertedSources = new Set(jsonlJsonFiles.map((file) => file.absolutePath));

  for (const file of files) {
    if (convertedSources.has(file.absolutePath)) {
      const csvRelative = replaceExtension(file.relativePath, '.csv');
      const csvPath = path.join(stageDir, csvRelative);
      convertedRows += convertJsonLinesToCsv(file.absolutePath, csvPath);
      effectiveExtensions.add('csv');
    } else {
      const dest = path.join(stageDir, file.relativePath);
      fs.mkdirSync(path.dirname(dest), { recursive: true });
      fs.copyFileSync(file.absolutePath, dest);
      effectiveExtensions.add(file.extension);
    }
  }

  return {
    dataset: stageDir,
    extensions: [...effectiveExtensions].sort(),
    diagnostics: [
      `normalized ${jsonlJsonFiles.length} .json JSONL file(s) to CSV (${convertedRows} row(s)) at ${stageDir}`,
    ],
  };
}

function normalizeExtension(value: string): string {
  return value.trim().toLowerCase().replace(/^\./, '');
}

const TABULAR_DISCOVERY_EXTENSIONS = new Set(['csv', 'tsv']);
// Mirror the enhancer's eval-set exclusion (enhance_for_copilot.ts datasetFiles)
// so detection never counts the operator-supplied ground-truth as data.
const EXCLUDED_DISCOVERY_DIR_NAMES = new Set(['evalset', 'eval-set', 'eval_set', 'workspace', 'node_modules', '.git']);
const EXCLUDED_DISCOVERY_FILE_PATTERN = /^eval[-_]?set/i;

/**
 * Detect the data-bearing file extensions actually present in a dataset,
 * mirroring the enhancer's eval-set exclusion rules. Returns a sorted list of
 * supported discovery extensions (empty if none / path unreadable).
 */
export function detectDatasetExtensions(datasetPath: string): string[] {
  const resolved = path.resolve(datasetPath);
  let stat: fs.Stats;
  try {
    stat = fs.statSync(resolved);
  } catch {
    return [];
  }
  const found = new Set<string>();
  const consider = (name: string): void => {
    if (EXCLUDED_DISCOVERY_FILE_PATTERN.test(name)) return;
    const ext = normalizeExtension(path.extname(name));
    if (ext && SUPPORTED_DISCOVERY_EXTENSIONS.has(ext)) found.add(ext);
  };
  if (stat.isFile()) {
    consider(path.basename(resolved));
    return [...found].sort();
  }
  const walk = (dir: string): void => {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (entry.name.startsWith('_') || EXCLUDED_DISCOVERY_DIR_NAMES.has(entry.name.toLowerCase())) continue;
        walk(path.join(dir, entry.name));
      } else if (entry.isFile()) {
        consider(entry.name);
      }
    }
  };
  walk(resolved);
  return [...found].sort();
}

/**
 * Choose the extensions to pass to the enhancer when the operator did not
 * specify any. Returns null to keep the enhancer's csv/tsv default — used when
 * tabular data files are present (preserving established behavior). Returns a
 * concrete list only when there are non-tabular data files and no tabular data
 * files, which is exactly the case the csv/tsv default silently drops (e.g. a
 * dataset of .json documents).
 */
export function autoEnhancerExtensions(datasetPath: string): string[] | null {
  const present = detectDatasetExtensions(datasetPath);
  if (present.length === 0) return null;
  if (present.some((e) => TABULAR_DISCOVERY_EXTENSIONS.has(e))) return null;
  return present;
}

function collectDatasetFiles(datasetPath: string, extensionFilter?: Set<string>): DatasetFile[] {
  const resolved = path.resolve(datasetPath);
  const stat = fs.statSync(resolved);
  const files: DatasetFile[] = [];

  const includeFile = (absolutePath: string, relativePath: string): void => {
    const extension = normalizeExtension(path.extname(absolutePath));
    if (!extension || !SUPPORTED_DISCOVERY_EXTENSIONS.has(extension)) return;
    if (extensionFilter && !extensionFilter.has(extension)) return;
    files.push({ absolutePath, relativePath, extension });
  };

  if (stat.isFile()) {
    includeFile(resolved, path.basename(resolved));
    return files;
  }

  const walk = (dir: string): void => {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
      } else if (entry.isFile()) {
        includeFile(fullPath, path.relative(resolved, fullPath));
      }
    }
  };

  walk(resolved);
  return files.sort((a, b) => a.relativePath.localeCompare(b.relativePath));
}

function isJsonLinesFile(filePath: string): boolean {
  const content = stripBom(fs.readFileSync(filePath, 'utf-8'));
  try {
    JSON.parse(content);
    return false;
  } catch {
    // Continue with JSONL detection below.
  }

  const lines = content.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  if (lines.length < 2) return false;

  for (const line of lines) {
    try {
      const parsed = JSON.parse(line);
      if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) return false;
    } catch {
      return false;
    }
  }

  return true;
}

function convertJsonLinesToCsv(inputPath: string, outputPath: string): number {
  const lines = stripBom(fs.readFileSync(inputPath, 'utf-8'))
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);
  const rows: Array<Record<string, unknown>> = [];
  const columns: string[] = [];
  const columnsSeen = new Set<string>();

  for (const line of lines) {
    const row = JSON.parse(line) as Record<string, unknown>;
    rows.push(row);
    for (const key of Object.keys(row)) {
      if (!columnsSeen.has(key)) {
        columnsSeen.add(key);
        columns.push(key);
      }
    }
  }

  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  const csv = [
    columns.map(csvEscape).join(','),
    ...rows.map((row) => columns.map((column) => csvEscape(stringifyCell(row[column]))).join(',')),
  ].join('\n');
  fs.writeFileSync(outputPath, `${csv}\n`, 'utf-8');
  return rows.length;
}

function stringifyCell(value: unknown): string {
  if (value === null || value === undefined) return '';
  if (typeof value === 'string') return value;
  if (typeof value === 'number' || typeof value === 'boolean') return String(value);
  return JSON.stringify(value);
}

function csvEscape(value: string): string {
  if (!/[",\r\n]/.test(value)) return value;
  return `"${value.replace(/"/g, '""')}"`;
}

function replaceExtension(relativePath: string, extension: string): string {
  const parsed = path.parse(relativePath);
  return path.join(parsed.dir, `${parsed.name}${extension}`);
}

function stripBom(value: string): string {
  return value.charCodeAt(0) === 0xfeff ? value.slice(1) : value;
}

export const _test = {
  prepareDatasetForWorkflow,
  isJsonLinesFile,
  convertJsonLinesToCsv,
};

File: winui3/src/Ccw.Core/Dataset/DatasetNormalization.cs
// Port of src/dataset-normalization.ts. Some source exports use newline-delimited
// JSON but keep a .json extension; EvalGen treats .json as a single JSON document.
// This module detects that disguised JSONL shape and converts to CSV so tabular
// records stay tabular, leaving other files untouched.

using System.IO;
using System.Text;
using System.Text.Json;

namespace Ccw.Core.Dataset;

public sealed record PreparedDataset
{
    public required string Dataset { get; init; }
    public IReadOnlyList<string>? Extensions { get; init; }
    public required IReadOnlyList<string> Diagnostics { get; init; }
}

public static class DatasetNormalization
{
    private static readonly HashSet<string> SupportedDiscoveryExtensions = new(StringComparer.Ordinal)
    {
        "csv", "tsv", "json", "jsonl", "xlsx", "xls", "docx", "pdf", "pptx", "txt", "md",
    };

    public static PreparedDataset PrepareDatasetForWorkflow(
        string datasetPath,
        string workspace,
        IReadOnlyList<string>? extensions = null)
    {
        var extensionFilter = extensions is { Count: > 0 }
            ? new HashSet<string>(extensions.Select(NormalizeExtension).Where(e => e.Length > 0), StringComparer.Ordinal)
            : null;
        var files = CollectDatasetFiles(datasetPath, extensionFilter);
        var jsonlJsonFiles = files
            .Where(f => f.Extension == "json" && IsJsonLinesFile(f.AbsolutePath))
            .ToList();

        if (jsonlJsonFiles.Count == 0)
        {
            return new PreparedDataset { Dataset = datasetPath, Extensions = extensions, Diagnostics = [] };
        }

        var stageDir = Path.Combine(workspace, "00-normalized-dataset");
        if (Directory.Exists(stageDir)) Directory.Delete(stageDir, recursive: true);
        Directory.CreateDirectory(stageDir);

        var effectiveExtensions = new HashSet<string>(StringComparer.Ordinal);
        var convertedRows = 0;
        var convertedSources = new HashSet<string>(jsonlJsonFiles.Select(f => f.AbsolutePath), StringComparer.OrdinalIgnoreCase);

        foreach (var file in files)
        {
            if (convertedSources.Contains(file.AbsolutePath))
            {
                var csvRelative = ReplaceExtension(file.RelativePath, ".csv");
                var csvPath = Path.Combine(stageDir, csvRelative);
                convertedRows += ConvertJsonLinesToCsv(file.AbsolutePath, csvPath);
                effectiveExtensions.Add("csv");
            }
            else
            {
                var dest = Path.Combine(stageDir, file.RelativePath);
                Directory.CreateDirectory(Path.GetDirectoryName(dest)!);
                File.Copy(file.AbsolutePath, dest, overwrite: true);
                effectiveExtensions.Add(file.Extension);
            }
        }

        return new PreparedDataset
        {
            Dataset = stageDir,
            Extensions = effectiveExtensions.OrderBy(x => x, StringComparer.Ordinal).ToList(),
            Diagnostics =
            [
                $"normalized {jsonlJsonFiles.Count} .json JSONL file(s) to CSV ({convertedRows} row(s)) at {stageDir}",
            ],
        };
    }

    internal sealed record DatasetFile(string AbsolutePath, string RelativePath, string Extension);

    private static string NormalizeExtension(string value) =>
        value.Trim().ToLowerInvariant().TrimStart('.');

    internal static List<DatasetFile> CollectDatasetFiles(string datasetPath, HashSet<string>? extensionFilter)
    {
        var resolved = Path.GetFullPath(datasetPath);
        var files = new List<DatasetFile>();

        void Include(string absolutePath, string relativePath)
        {
            var ext = NormalizeExtension(Path.GetExtension(absolutePath));
            if (string.IsNullOrEmpty(ext) || !SupportedDiscoveryExtensions.Contains(ext)) return;
            if (extensionFilter is not null && !extensionFilter.Contains(ext)) return;
            files.Add(new DatasetFile(absolutePath, relativePath, ext));
        }

        if (File.Exists(resolved))
        {
            Include(resolved, Path.GetFileName(resolved));
            return files;
        }

        if (!Directory.Exists(resolved)) return files;

        void Walk(string dir)
        {
            foreach (var entry in Directory.GetFileSystemEntries(dir))
            {
                if (Directory.Exists(entry))
                {
                    Walk(entry);
                }
                else if (File.Exists(entry))
                {
                    var rel = Path.GetRelativePath(resolved, entry);
                    Include(entry, rel);
                }
            }
        }

        Walk(resolved);
        return files.OrderBy(f => f.RelativePath, StringComparer.Ordinal).ToList();
    }

    public static bool IsJsonLinesFile(string filePath)
    {
        var content = StripBom(File.ReadAllText(filePath));
        try
        {
            using var _ = JsonDocument.Parse(content);
            return false;
        }
        catch (JsonException)
        {
            // continue with JSONL detection
        }

        var lines = content.Split('\n', '\r')
            .Select(l => l.Trim())
            .Where(l => l.Length > 0)
            .ToList();
        if (lines.Count < 2) return false;

        foreach (var line in lines)
        {
            try
            {
                using var doc = JsonDocument.Parse(line);
                if (doc.RootElement.ValueKind != JsonValueKind.Object) return false;
            }
            catch (JsonException) { return false; }
        }

        return true;
    }

    public static int ConvertJsonLinesToCsv(string inputPath, string outputPath)
    {
        var content = StripBom(File.ReadAllText(inputPath));
        var lines = content.Split('\n', '\r')
            .Select(l => l.Trim())
            .Where(l => l.Length > 0)
            .ToList();

        var rows = new List<Dictionary<string, JsonElement>>();
        var columns = new List<string>();
        var columnsSeen = new HashSet<string>(StringComparer.Ordinal);

        foreach (var line in lines)
        {
            using var doc = JsonDocument.Parse(line);
            if (doc.RootElement.ValueKind != JsonValueKind.Object) continue;
            var row = new Dictionary<string, JsonElement>(StringComparer.Ordinal);
            foreach (var prop in doc.RootElement.EnumerateObject())
            {
                row[prop.Name] = prop.Value.Clone();
                if (columnsSeen.Add(prop.Name)) columns.Add(prop.Name);
            }
            rows.Add(row);
        }

        Directory.CreateDirectory(Path.GetDirectoryName(outputPath)!);
        var sb = new StringBuilder();
        sb.Append(string.Join(",", columns.Select(CsvEscape)));
        foreach (var row in rows)
        {
            sb.Append('\n');
            sb.Append(string.Join(",", columns.Select(c =>
                CsvEscape(row.TryGetValue(c, out var v) ? StringifyCell(v) : string.Empty))));
        }
        sb.Append('\n');
        File.WriteAllText(outputPath, sb.ToString());
        return rows.Count;
    }

    private static string StringifyCell(JsonElement v) => v.ValueKind switch
    {
        JsonValueKind.Null => string.Empty,
        JsonValueKind.Undefined => string.Empty,
        JsonValueKind.String => v.GetString() ?? string.Empty,
        JsonValueKind.Number => v.GetRawText(),
        JsonValueKind.True => "true",
        JsonValueKind.False => "false",
        _ => v.GetRawText(),
    };

    private static string CsvEscape(string value)
    {
        if (value.IndexOfAny(['"', ',', '\r', '\n']) < 0) return value;
        return "\"" + value.Replace("\"", "\"\"", StringComparison.Ordinal) + "\"";
    }

    private static string ReplaceExtension(string relativePath, string extension)
    {
        var dir = Path.GetDirectoryName(relativePath) ?? string.Empty;
        var name = Path.GetFileNameWithoutExtension(relativePath);
        return string.IsNullOrEmpty(dir) ? name + extension : Path.Combine(dir, name + extension);
    }

    private static string StripBom(string value) =>
        value.Length > 0 && value[0] == '\uFEFF' ? value[1..] : value;
}
